libr/util/big-ssl.c

```diff
@@ -1,89 +1,122 @@
+    if (v < 0) {
+        BN_set_word (b, -v);
+        BN_set_negative (b, v);
+    } else {
+        BN_set_word (b, v);
+    }
 }
 
-R_API void r_big_set(RNumBig *a, RNumBig *b) {
-	BN_copy (a, b);
+R_API signed long int r_big_to_int(RNumBig *b) {
+    signed long int res;
+    res = BN_get_word (b);
+    res *= (BN_is_negative (b) ? -1 : 1);
 }
 
-R_API void r_big_set_st(RNumBig *n, int v) {
-	BN_set_word (n, v);
+R_API void r_big_from_hexstr (RNumBig *b, const char *str) {
+    BN_hex2bn (&b, str);
 }
 
-R_API void r_big_set_st64(RNumBig *n, st64 v) {
-	BN_set_word (n, v);
+R_API char *r_big_to_hexstr (RNumBig *b) {
+    return BN_bn2hex (b);
 }
 
-/* c = a [+*-/] b; */
-R_API void r_big_add (RNumBig *c, RNumBig *a, RNumBig *b) {
-	BN_add (c, a, b);
+R_API void r_big_assign (RNumBig *dst, RNumBig *src) {
+    BN_copy (dst, src);
 }
 
-R_API void r_big_sub(RNumBig *c, RNumBig *a, RNumBig *b) {
-	BN_sub (c, a, b);
+R_API void r_big_add (RNumBig *c, RNumBig *a, RNumBig *b) {
+    BN_add (c, a, b);
 }
 
-R_API int r_big_cmp(RNumBig *a, RNumBig *b) {
-	return BN_ucmp (a, b);
+R_API void r_big_sub (RNumBig *c, RNumBig *a, RNumBig *b) {
+    BN_sub (c, a, b);
 }
 
-R_API int r_big_cmp_st(RNumBig *n, int v) {
-	return BN_cmp (a, b);
+R_API void r_big_mul (RNumBig *c, RNumBig *a, RNumBig *b) {
+    BN_CTX *bn_ctx = BN_CTX_new();
+    BN_mul (c, a, b, bn_ctx);
+    BN_CTX_free(bn_ctx);
 }
 
-/* multiply n by 10^d */
-R_API void r_big_shift(RNumBig *n, int d) {
-	RNumBig a;
-	BN_copy (&a, n);
-	BN_lshift, (n, a, d);
+R_API void r_big_div (RNumBig *c, RNumBig *a, RNumBig *b) {
+    BN_CTX *bn_ctx = BN_CTX_new();
+    BN_div (c, NULL, a, b, bn_ctx);
+    BN_CTX_free(bn_ctx);
 }
 
-R_API void r_big_mul (RNumBig *c, RNumBig *a, RNumBig *b) {
-	/* TODO: last parameter must be BN_CTX */
-	BN_mul (c, a, b, NULL);
+R_API void r_big_mod (RNumBig *c, RNumBig *a, RNumBig *b) {
+    BN_CTX *bn_ctx = BN_CTX_new();
+    BN_mod (c, a, b, bn_ctx);
+    BN_CTX_free(bn_ctx);
 }
 
-R_API void r_big_mul_ut (RNumBig *c, RNumBig *a, ut32 b) {
-	RNumBig *b = BN_new ();
-	BN_set_word (b, b);
-	r_big_mul (c, a, &b);
-	BN_free (b);
+R_API void r_big_divmod (RNumBig *c, RNumBig *d, RNumBig *a, RNumBig *b) {
+    BN_CTX *bn_ctx = BN_CTX_new();
+    BN_div (c, d, a, b, bn_ctx);
+    BN_CTX_free(bn_ctx);
 }
 
-R_API void r_big_div(RNumBig *c, RNumBig *a, RNumBig *b) {
-	BN_div (c, NULL, a, b);
-}
+R_API void r_big_and (RNumBig *c, RNumBig *a, RNumBig *b) {
 
-R_API void r_big_div_ut(RNumBig *c, RNumBig *a, ut32 b) {
-	eprintf (__FUNCTION__": TODO\n");
-//	mpz_divexact_ui (*c, *a, b);
-}
+} 
 
-R_API int r_big_divisible_ut(RNumBig *n, ut32 v) {
-	eprintf (__FUNCTION__": TODO\n");
-//	return mpz_divisible_ui_p (*n, v);
-}
+R_API void r_big_or (RNumBig *c, RNumBig *a, RNumBig *b) {
 
-R_API void r_big_mod(RNumBig *c, RNumBig *a, RNumBig *b) {
-	BN_div (NULL, c, a, b);
-}
+}  
+
+R_API void r_big_xor (RNumBig *c, RNumBig *a, RNumBig *b) {
+
```

Comment:
xor can be implemented like this: `(a XOR b) = ((a OR b) - (a AND b))`

Reply:
I know, but how to impl `OR` & `AND`?